src/main/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentBuildWrapper.java

```diff
@@ -296,95 +293,20 @@ public String getDisplayName() {
-                add(sshUserPrivateKey);
-            }
-        }
+        private final FilePath workspace;
 
-        @Deprecated
-        public SSHAgentEnvironment(Launcher launcher, final BuildListener listener) throws Throwable {
-            this(launcher, listener, (FilePath) null);
-        }
+        private final BuildListener listener;
 
-        /**
-         * Construct the environment and initialize on the remote node.
-         *
-         * @param launcher           the launcher for the remote node.
-         * @param listener           the listener for reporting progress.
-         * @throws Throwable if things go wrong.
-         * @since 1.9
-         */
-        public SSHAgentEnvironment(Launcher launcher, BuildListener listener, @CheckForNull FilePath workspace) throws Throwable {
-            RemoteAgent agent = null;
+        SSHAgentEnvironment(Launcher launcher, BuildListener listener, FilePath workspace) throws Throwable {
             this.launcher = launcher;
+            this.workspace = Objects.requireNonNull(workspace);
             this.listener = listener;
             listener.getLogger().println("[ssh-agent] Looking for ssh-agent implementation...");
-            Map<String, Throwable> faults = new LinkedHashMap<>();
-            for (RemoteAgentFactory factory : Jenkins.get().getExtensionList(RemoteAgentFactory.class)) {
-                if (factory.isSupported(launcher, listener)) {
-                    try {
-                        listener.getLogger().println("[ssh-agent]   " + factory.getDisplayName());
-                        agent = factory.start(new SingletonLauncherProvider(launcher), listener,
-                                workspace != null ? SSHAgentStepExecution.tempDir(workspace) : null);
-                        break;
-                    } catch (Throwable t) {
-                        faults.put(factory.getDisplayName(), t);
-                    }
-                }
-            }
-            if (agent == null) {
-                listener.getLogger().println("[ssh-agent] FATAL: Could not find a suitable ssh-agent provider");
-                listener.getLogger().println("[ssh-agent] Diagnostic report");
-                for (Map.Entry<String, Throwable> fault : faults.entrySet()) {
-                    listener.getLogger().println("[ssh-agent] * " + fault.getKey());
-                    StringWriter sw = new StringWriter();
-                    fault.getValue().printStackTrace(new PrintWriter(sw));
-                    for (String line : StringUtils.split(sw.toString(), "\n")) {
-                        listener.getLogger().println("[ssh-agent]     " + line);
-                    }
-                }
-                throw new RuntimeException("[ssh-agent] Could not find a suitable ssh-agent provider.");
-            }
-            this.agent = agent;
+            agent = new ExecRemoteAgent(launcher, listener);
```

Comment:
When there is no list of factories to consult, the code is simpler and the error handling should be as well: either running `ssh-agent` works, or it does not.